### PR TITLE
docs(ds): Update IMPLEMENTATION.md

### PR DESCRIPTION
**Pull request by ieQu1**
Fixes <issue-or-jira-number>

### <samp>🤖 Generated by Copilot at e0c11f5</samp>

This change improves the documentation of the `emqx_durable_storage` feature by adding more details, clarifying the design and workflow, and fixing some formatting and typos in the `IMPLEMENTATION.md` file.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_durable_storage/IMPLEMENTATION.md [hunk 1]
```diff
@@ -3,9 +3,17 @@
 In the logic layer we don't speak about replication.
 This is because we could use an external DB with its own replication logic.
 
-On the other hand, we introduce notion of shard right here at the logic.
+On the other hand, we introduce notion of shard right here at the logic layer.
 This is because shared subscription logic needs to be aware of it to some extend, as it has to split work between subscribers somehow.
 
+
+# Modus operandi
```
Comment (id) [inline apps/emqx_durable_storage/IMPLEMENTATION.md:10]: Suggest to add 'First delivery' section with the following content (below). Based on my meetings notes from the first part of the session about first delivery scope, processed with ChatGPT (as you could tell).

> The recent efforts to integrate RocksDB into our system are primarily aimed at improving the existing mechanism, which is based on ETS tables. The shortcoming of the current mechanism lies in its incapability to preserve messages when a node goes offline, as it is entirely RAM-based. Additionally, its high demand for RAM significantly restricts scalability.
> In the past, we made attempts to use Mnesia as a possible alternative. However, the performance test did not yield promising results, leading us to explore other alternatives. Our ultimate objective, however, is not to devise a flawless solution akin to systems such as Kafka. Instead, our focus is on delivering an improved solution compared to previous iterations.
> The key features we aim to implement include durability, scalability, and an adequate speed. The first phase of the delivery will not incorporate a replication feature, implying that there will only be a single copy of messages in the cluster. Consequently, if a node goes offline, we would temporarily lose access to the messages stored on that node. However, these messages will be recoverable once the node is back online.
While the initial version of this integration may not feature replication, we are designing it with a foresight for future replication capabilities. We are optimistic that this new integration will significantly elevate the efficiency and reliability of our system.

**Review by keynslug — APPROVED** [review]

**Review by id — DISMISSED** [review]

File: apps/emqx_durable_storage/IMPLEMENTATION.md (hunk 1 above)
Comment (id) [reply apps/emqx_durable_storage/IMPLEMENTATION.md:10]: nvm, added to Confluence

File: apps/emqx_durable_storage/IMPLEMENTATION.md [hunk 2]
```diff
@@ -3,9 +3,17 @@
...
+
+1. Create a draft implementation of a milestone
+2. Performance
```
Comment (thalesmg) [inline apps/emqx_durable_storage/IMPLEMENTATION.md:13]:
```suggestion
2. Test performance
```
?

File: apps/emqx_durable_storage/IMPLEMENTATION.md [hunk 3]
```diff
@@ -40,36 +51,111 @@ Data format:
...
-- Need pushback logic
+- Needs pushback logic
 - It's not entirely justified when working with external DB that may not provide streaming API
 
 ## Pull pros:
 - No need for pushback: client advances iterators at its own tempo
--
+
+## Pull cons:
+- 2 messages need to be sent over network for each batch being replayed.
```
Comment (thalesmg) [inline apps/emqx_durable_storage/IMPLEMENTATION.md:95]: Q: which messages?  One requesting the iterator to advance, and its response with the payload?

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_durable_storage/IMPLEMENTATION.md (hunk 3 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/IMPLEMENTATION.md:95]: Yes, response and payload.

**Review by id — APPROVED** [review]

**Review by thalesmg — APPROVED** [review]

**Pull request merged**